Implemented history resource

om-sharma opened the pull request:
Added history page and a resource to return all the versions of an entry.

om-sharma committed:
Added history page and a resource to return all the versions of an entry.

om-sharma committed:
Removed string interpolation and using servletresponse link header to get history link

philandstuff commented on src/main/resources/templates/entry.html:
```diff
@@ -47,6 +47,12 @@ <h1 class="entry-header" th:text="${record.primaryKey}">Entry Name Header</h1>
 
+<div class="row">
+    <div class="small-12 large-6 columns">
+        <p>View all versions of this record <a th:href="${@uk.gov.register.presentation.HtmlViewSupport@extractLinkFromLinkHeader(#ctx.httpServletResponse.getHeader('Link'))}">here</a></p>
```
extracting the link from the link header feels fragile (because we might later add another link rel in another header) and wrong (because the view should be able to provide the link directly)

om-sharma replied:
After discussion we thought that it is ok to keep this way for now.

philandstuff commented on src/main/java/uk/gov/register/presentation/resource/HistoryResource.java:
```diff
@@ -0,0 +1,61 @@
+        }
+    }
+
+    private static class VersionJsonSerializer extends JsonSerializer<ListVersionView> {
```
This class name suggests it serializes `Version`, not `ListVersionView`...

om-sharma replied:
Renamed to ListVersionViewJsonSerializer in next commit.

philandstuff commented on src/main/java/uk/gov/register/presentation/resource/SearchResource.java:
```diff
@@ -54,6 +54,7 @@ public SingleResultView findByPrimaryKey(@PathParam("primaryKey") String key, @P
         if (key.equals(registerPrimaryKey)) {
             Optional<DbRecord> record = queryDAO.findByKeyValue(key, value);
             if (record.isPresent()) {
+                setResponseHeader(key, value);
```
this method name is unclear

om-sharma replied:
Renamed to setVersionHistoryLinkHeader in next commit.

philandstuff commented on src/test/java/uk/gov/register/presentation/resource/HistoryResourceTest.java:
```diff
@@ -0,0 +1,28 @@
+        try {
+            resource.history("someOtherKey", "value");
+            fail("Must fail");
+        } catch (NotFoundException e) {
```
you can use `@Test(expected = NotFoundException.class)` to avoid the try/catch block here

om-sharma replied:
Should have try/catch block to confirm that the exception is thrown from the method call only.

om-sharma committed:
Added quotes in link header rel value
Renamed class VersionJsonSerializer to ListVersionViewJsonSerializer
Renamed method setResponseHeader to be more specific as setVersionHistoryLinkHeader
